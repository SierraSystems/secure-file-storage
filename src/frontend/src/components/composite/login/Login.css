.outline {
  background-color: rgba(239, 243, 252, 1);
  box-sizing: border-box;
  border-width: 1px;
  border-style: solid;
  border-color: rgba(190, 206, 243, 1);
  border-radius: 6px;
  width: 36%;
  min-height: 70px;
  font-weight: 700;
  font-style: normal;
  font-size: 16px;
  color: #003366;
  text-align: left;
  margin-top: 20px;
  line-height: 20px;
  padding-top: 15px;
  padding-bottom: 12px;
  margin: auto;
}

@media only screen and (max-device-width: 480px) {
  .outline {
    width: 80%;
    margin-bottom: 50px;
  }
}

.center-align {
  width: fit-content;
  margin: auto;
}

.loader-align {
  width: 100%;
  text-align: center;
  display: inline-block;
}
