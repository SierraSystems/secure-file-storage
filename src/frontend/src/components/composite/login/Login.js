import React, { useState, useEffect } from "react";
import { useLocation, useHistory } from "react-router-dom";
import axios from "axios";
import queryString from "query-string";
import "./Login.css";
import "bootstrap/dist/css/bootstrap.css";
import { BcscButton } from "../../base/bcscbutton/BcscButton";
import Loader from "../../base/loader/Loader";
import Uploader from "../../base/uploader/Uploader";
import { isAuthenticated } from "../../../AuthenticationHelper";
import UserInfo from "../../base/userinfo/UserInfo";

let loginUrl;
const basicAuth = {
  auth: {
    username: sessionStorage.getItem("username"),
    password: sessionStorage.getItem("password")
  }
};

const authMessage = message => (
  <div className="spacing">
    <p className="text head">{message}</p>
  </div>
);

export const redirectUser = () => {
  window.open(loginUrl, "_self");
};

export const generateRedirectUrl = () => {
  axios
    .get("http://localhost:8080/oauth/getBCSCUrl", basicAuth)
    .then(response => {
      loginUrl = response.data;
    })
    .catch(() => {});

  return loginUrl;
};

export const loginUser = (code, setAuthed, setMessage, path, history) => {
  axios
    .get(`http://localhost:8080/oauth/login?code=${code}`, basicAuth)
    .then(response => {
      sessionStorage.setItem("jwt", response.data);
      if (isAuthenticated()) {
        setAuthed(true);
        setMessage(
          "You have been securely authenticated. Your information is presented below. Enjoy using the service!"
        );
      }
    })
    .catch(() => {
      sessionStorage.clear();
      history.push(path);
    });
};

export const Login = () => {
  const [authed, setAuthed] = useState(false);
  const [message, setMessage] = useState(
    "Please login using your BC Services Card to use the service. Click the button below to log in."
  );
  const history = useHistory();
  const location = useLocation();
  const urlParam = queryString.parse(location.search);
  const { code } = urlParam;

  useEffect(() => {
    if (code && !authed) {
      loginUser(code, setAuthed, setMessage, location.pathname, history);
    } else {
      loginUrl = generateRedirectUrl();
    }
  });

  return (
    <>
      {!code && (
        <>
          {authMessage(message)}
          <br />
          <BcscButton onClick={redirectUser} />
        </>
      )}
      {code && !authed && <Loader />}
      {authed && (
        <>
          {authMessage(message)}
          <br />
          <UserInfo />
          <br />
          <Uploader />
        </>
      )}
    </>
  );
};
