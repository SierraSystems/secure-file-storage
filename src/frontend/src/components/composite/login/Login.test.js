import { redirectUser, loginUser, generateRedirectUrl } from "./Login";
import { createMemoryHistory } from "history";
import axios from "axios";
import { wait } from "@testing-library/react";

const MockAdapter = require("axios-mock-adapter");

window.open = jest.fn();

describe("Login", () => {
  let mock;
  const code = "abc123";
  const setAuthed = jest.fn();
  const setMessage = jest.fn();
  const path = "/securefileupload/applicationform";
  const history = createMemoryHistory();

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  test("Redirect user function opens the redirect login url", () => {
    redirectUser();

    expect(window.open).toHaveBeenCalledTimes(1);
  });

  test("Login user function clears session storage and takes you back to unauthed homepage on fail", async () => {
    sessionStorage.setItem("jwt", "token");

    expect(sessionStorage.getItem("jwt")).toBeTruthy();

    mock.onGet(`http://localhost:8080/oauth/login?code=${code}`).reply(400);

    loginUser(code, setAuthed, setMessage, path, history);

    await wait(() => {
      expect(sessionStorage.getItem("jwt")).toBeFalsy();
    });
  });

  test("Login user function sets jwt token on success", async () => {
    mock
      .onGet(`http://localhost:8080/oauth/login?code=${code}`)
      .reply(200, "token123");

    loginUser(code, setAuthed, setMessage, path, history);

    await wait(() => {
      expect(sessionStorage.getItem("jwt")).toEqual("token123");
    });
  });

  test("Generate redirect url function does not set loginUrl when unsuccessful", async () => {
    mock.onGet("http://localhost:8080/oauth/getBCSCUrl").reply(400);

    const result = generateRedirectUrl();

    await wait(() => {
      expect(result).toBeFalsy();
    });
  });
});
