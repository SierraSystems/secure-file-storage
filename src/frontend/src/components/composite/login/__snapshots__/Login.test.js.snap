// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login authMessageAndComponent matches a snapshot when authed 1`] = `
Array [
  <br />,
  <div
    className="spacing"
  >
    <p
      className="head"
    >
      <b>
        Name: 
      </b>
    </p>
    <p
      className="head"
    >
      <b>
        Birthdate: 
      </b>
    </p>
    <p
      className="head"
    >
      <b>
        Gender: 
      </b>
    </p>
    <p
      className="head"
    >
      <b>
        Address: 
      </b>
    </p>
  </div>,
  <br />,
  <div
    className="spacing custom-width"
  >
    <div
      className="filepond--wrapper"
    >
      <input
        multiple={true}
        type="file"
      />
    </div>
  </div>,
]
`;

exports[`Login authMessageAndComponent matches a snapshot when unauthed 1`] = `
Array [
  <br />,
  <div
    className="outline"
  >
    <div
      className="center-align"
    >
      <button
        className="login_bcsc btn"
        onClick={[MockFunction]}
        type="button"
      >
        Login with a BC Services Card
      </button>
    </div>
  </div>,
]
`;
